src/Microsoft.AspNet.Hosting/Internal/RequestServicesContainerFeature.cs

```diff
@@ -0,0 +1,55 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNet.Http.Features.Internal;
+using Microsoft.Framework.DependencyInjection;
+
+namespace Microsoft.AspNet.Hosting.Internal
+{
+    public class RequestServicesFeature : IServiceProvidersFeature, IDisposable
+    {
+        private IServiceProvider _requestServices;
+        private IServiceScope _scope;
+        private bool _requestServicesSet;
+
+        public RequestServicesFeature(IServiceProvider applicationServices)
+        {
+            if (applicationServices == null)
+            {
+                throw new ArgumentNullException(nameof(applicationServices));
+            }
+
+            ApplicationServices = applicationServices;
+        }
+
+        public IServiceProvider ApplicationServices { get; set; }
+
+        public IServiceProvider RequestServices
+        {
+            get
+            {
+                if (!_requestServicesSet)
+                {
+                    _scope = ApplicationServices.GetRequiredService<IServiceScopeFactory>().CreateScope();
```

Comment:
Yeah looks like it.  Do we want to throw an exception if you try to set it to null, or just treat that as no-op and basically do nothing and leave RequestServices null if you set ApplicationServices to null?